### DISCUSSION TITLE
Support for testing downloads? (assert_download)

### DISCUSSION THREAD

**Discussion by RobinBoers** [keathley/wallaby#703] [Ideas]
I'm building an app that has file sharing with end-to-end encryption. The files get decrypted in the frontend and are then downloaded via a blob URL. 

I looked at the docs and there doesn't seem to be a way to test file downloads. I thought it would be cool to be able to test file downloads using something like `assert_download`?

**Comment by agatheblues** [comment]
I gave this a quick try with:

```
  @download_dir "path/to/tmp"
  @sessions [
      [capabilities: %{
        chromeOptions: %{
          prefs: %{
            "download.default_directory": @download_dir
          }
        }
      }]
    ]
   setup do
      on_exit(fn -> File.rm_rf!(@download_dir)
   end

    feature "downloads things" do
         # click the download button
          ...
         @download_dir |> Path.join("data.zip") |> File.read!()
         #  assert on file(s)
    end
```

But it turned out to be very flaky. Sometimes the file is there sometimes it isn't. I tried another attempt where I would make sure the directory exists / or else create it, + remove the download directory in the `on_exit` callback but with no luck.

Also tried to use `System.tmp_dir!()`, same story.

Finally I tried to use the `@tag :tmp_dir` of ExUnit but I don't have access to the actual `tmp_dir` to initialize the `@sessions` with the right directory.

So, didn't get to make this work 🤷‍♀️ I think the flakiness comes from _"ChromeDriver does not automatically wait for download to complete. If you call driver.quit() too soon, Chrome might terminate before the download has finished."_ (https://chromedriver.chromium.org/capabilities) and I don't know how to go around this - would even think there's nothing to do about this.

Alternatively in our other e2e tests, when the download is served by the controller, we read the `href` in the link and manually make a GET request to that url with tesla. Then we can do assertions on the body. But in the case I was trying to solve we are sending the data directly from liveview to the client as a blob, like you described so that little trick doesn't work :( 

Another solution I'd think of is to leave the link with blob URL on the page (in our case we append it / remove it on the fly) and parse the data in the `href` attribute to make assertions.
